weave/trace_server/refs_internal.py

```diff
@@ -17,16 +17,82 @@
 OBJECT_ATTR_EDGE_NAME = "attr"
 TABLE_ROW_ID_EDGE_NAME = "id"
 
+valid_edge_names = (
+    DICT_KEY_EDGE_NAME,
+    LIST_INDEX_EDGE_NAME,
+    OBJECT_ATTR_EDGE_NAME,
+    TABLE_ROW_ID_EDGE_NAME,
+)
+
 
 class InvalidInternalRef(ValueError):
     pass
 
 
+def quote_select(s: str, quote_chars: tuple[str, ...] = ("/", ":", "%")) -> str:
+    """We don't need to quote every single character, rather
+    we just need to quote the characters that are used as
+    delimiters in the URI. Right now, we only use "/", ":", and "%". Moreover,
+    the only user-controlled fields are the object name and the extra fields.
+
+    All other fields are generated by the system and are safe for parsing.
+    """
+    # We have to be careful here since quoting creates new "%" characters.
+    # We don't want to double-quote characters.
+    if "%" in quote_chars and quote_chars[-1] != "%":
+        raise ValueError("Quoting '%' must be the last character in the list.")
+
+    for c in quote_chars:
+        s = s.replace(c, urllib.parse.quote(c))
+
+    return s
+
+
+def validate_extra(extra: list[str]) -> None:
+    if len(extra) % 2 != 0:
+        raise InvalidInternalRef("Extra fields must be key-value pairs.")
+    for i, e in enumerate(extra):
+        if i % 2 == 0:
```

Comment:
What does this branching mean?  It's not immediately obvious on reading

Reply:
i'll make it more clear. basically extra is always a series of `edge/value` pairs, where edge is a set of known edge types, and value is user-defined